sdks/java/io/hadoop/input-format/src/test/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIOIT.java

```diff
@@ -0,0 +1,133 @@
+import org.apache.beam.sdk.transforms.Reshuffle;
+import org.apache.beam.sdk.transforms.Values;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.mapreduce.InputFormat;
+import org.apache.hadoop.mapreduce.lib.db.DBConfiguration;
+import org.apache.hadoop.mapreduce.lib.db.DBInputFormat;
+import org.apache.hadoop.mapreduce.lib.db.DBWritable;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Rule;
+import org.junit.Test;
+import org.postgresql.ds.PGSimpleDataSource;
+
+/**
+ * IOIT for {@link org.apache.beam.sdk.io.hadoop.inputformat.HadoopInputFormatIO}.
+ */
+public class HadoopInputFormatIOIT {
+
+  private static PGSimpleDataSource dataSource;
+  private static Long numberOfRows;
+  private static String tableName;
+  private static SerializableConfiguration hadoopConfiguration;
+
+  @Rule
+  public TestPipeline writePipeline = TestPipeline.create();
+
+  @Rule
+  public TestPipeline readPipeline = TestPipeline.create();
+
+  @BeforeClass
+  public static void setUp() throws SQLException {
+    PipelineOptionsFactory.register(IOTestPipelineOptions.class);
+    IOTestPipelineOptions options = TestPipeline.testingPipelineOptions()
+        .as(IOTestPipelineOptions.class);
+
+    dataSource = DatabaseTestHelper.getPostgresDataSource(options);
+    numberOfRows = options.getNumberOfRecords();
+    tableName = DatabaseTestHelper.getTestTableName("HadoopInputFormatIOIT");
+
+    DatabaseTestHelper.createDataTable(dataSource, tableName);
+    setupHadoopConfiguration(options);
+  }
+
+  private static void setupHadoopConfiguration(IOTestPipelineOptions options) {
+    Configuration conf = new Configuration();
+    DBConfiguration.configureDB(
+        conf,
+        "org.postgresql.Driver",
+        DatabaseTestHelper.getPostgresDBUrl(options),
+        options.getPostgresUsername(),
+        options.getPostgresPassword()
+    );
+    conf.set(DBConfiguration.INPUT_TABLE_NAME_PROPERTY, tableName);
+    conf.setStrings(DBConfiguration.INPUT_FIELD_NAMES_PROPERTY, "id", "name");
+    conf.setClass(DBConfiguration.INPUT_CLASS_PROPERTY, TestRowDBWritable.class, DBWritable.class);
+
+    conf.setClass("key.class", LongWritable.class, Object.class);
+    conf.setClass("value.class", TestRowDBWritable.class, Object.class);
+    conf.setClass("mapreduce.job.inputformat.class", DBInputFormat.class, InputFormat.class);
+
+    hadoopConfiguration = new SerializableConfiguration(conf);
+  }
+
+  @AfterClass
+  public static void tearDown() throws SQLException {
+    DatabaseTestHelper.cleanUpDataTable(dataSource, tableName);
+  }
+
+  @Test
+  public void writeThenReadUsingDBInputFormat() {
+    writePipeline.apply("Generate sequence", GenerateSequence.from(0).to(numberOfRows))
+        .apply("Produce db rows", ParDo.of(new DeterministicallyConstructTestRowFn()))
+        .apply(JdbcIO.<TestRow>write()
```

Comment:
<!--thread_id:cc_159567128_t; commit:138b0e00d1ea86cd27700083ee657e8c5abae4a6; resolved:1-->
<!--section:context-quote-->
> **chamikaramj** wrote:
> Writing using JDBCIO is fine. Probably your pipeline is slow since you do all the writing From a single worker. Consider creating a PCollection of seed objects (splits) followed by a Reshuffle() followed by writing.

<!--section:body-->
Done.